apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,122 +243,699 @@ terminate(_Reason, _St) ->
+                    {R, C2};
+                #{ locked_until := LockedUntil } ->
+                    Min = maps:get(min_depth, R, 0),
+                    CheckAt = LockedUntil + Min,
+                    {R#{ locked_until    => LockedUntil
+                       , check_at_height => CheckAt }, C2};
+                undefined ->
+                    %% Set closing time to current top height, wait for min_depth
+                    Min = maps:get(min_depth, R, 0),
+                    {R#{ locked_until    => TopHeight
+                       , check_at_height => TopHeight + Min }, C2}
+            end;
+       true ->
+            %% Assume a close watcher is never started before funding_locked
+            %% If the channel state is undefined, assume that the channel has
+            %% been deleted. A corner case is that it re-appears after e.g. a
+            %% fork switch, but with a suitable min_depth value, it should be ok
+            %% to give up after TopHeight + MinDepth
+            {TopHeight, C1} = top_height(C),
+            Min = maps:get(min_depth, R, 0),
+            {R#{ locked_until    => TopHeight
+               , check_at_height => TopHeight + Min }, C1}
     end;
-check_status_(#{mode := watch} = R, St, C) ->
-    watch_for_channel_closing(R, St, C);
-check_status_(#{mode := tx_hash, tx_hash := TxHash, parent := Parent,
-                type := Type, min_depth := MinDepth} = R,
+check_req(#{mode := unlock} = R, #st{chan_id = ChId, chan_vsn = Vsn} = St, C) ->
+    if Vsn =/= undefined ->
+            lager:debug("checking", []),
+            case get_basic_ch_status_(ChId, C) of
+                {#{locked_until := 0} = Ch, C1} ->
+                    lager:debug("locked_until = 0", []),
+                    report_channel_unlocked(R, Ch, St, C1),
+                    {done, C1};
+                {#{locked_until := LockedUntil} = Ch, C1} ->
+                    {Height, C2} = top_height(C1),
+                    lager:debug("LockedUntil = ~p, Height = ~p", [LockedUntil, Height]),
+                    case LockedUntil < Height of
+                        true  ->
+                            lager:debug("locked_until expired", []),
+                            report_channel_unlocked(R, Ch, St, C2),
+                            {done, C2};
+                        false ->
+                            lager:debug("still locked", []),
+                            {R#{check_at_height => LockedUntil + 1}, C2}
+                    end;
+                {undefined, C1} ->
+                    lager:debug("couldn't find channel", []),
+                    {done, C1}
+            end;
+       true ->
+            lager:debug("Vsn = undefined", []),
+            {R, C}
+    end;
+check_req(#{mode := watch} = R, #st{chan_id = ChId} = St,
+          #{scenario := Scenario, top_hash := Hash } = C)
+  when ?IS_SCENARIO(Scenario) ->
+    case Scenario of
+        {has_tx, {_, #{ type    := TxType
```

Comment:
This assumes that second element in tuple that is the entry of the `aesc_window` (log) data structure has a `type` key.

Reply:
I see that only entries that satisfy filter `[{block_hash, BHash}]` end up here. That does not solve problem.